systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m

```diff
@@ -1136,31 +1148,36 @@
       %
       % See also RigidBodyGeometry/getTerrainContactPoints,
       % RigidBodyManipulator/terrainContactPositions
-      if nargin < 2
-        body_idx = 2:obj.getNumBodies(); % World-fixed objects can't collide
-                                         % with the terrain
-      end
-      if nargin >= 3
-        if all(cellfun(@ischar,contact_groups))
-          contact_groups = {contact_groups};
-        end
-        if numel(contact_groups) == 1
-          contact_groups = repmat(contact_groups,size(body_idx));
-        else
-          sizecheck(contact_groups,size(body_idx));
+      checkDirty(obj);
+      if nargin == 1 && ~isempty(obj.cached_terrain_contact_points_struct)
+        terrain_contact_point_struct = obj.cached_terrain_contact_points_struct;
+      else
+        if nargin < 2
+          body_idx = 2:obj.getNumBodies(); % World-fixed objects can't collide
+          % with the terrain
         end
-      end
-      terrain_contact_point_struct = struct('pts',{},'idx',{});
-      for i = 1:length(body_idx)
-        bi=body_idx(i);
-        if bi ~= 1
-          if nargin < 3
-            pts = getTerrainContactPoints(obj.body(bi));
+        if nargin >= 3
+          if all(cellfun(@ischar,contact_groups))
+            contact_groups = {contact_groups};
+          end
+          if numel(contact_groups) == 1
+            contact_groups = repmat(contact_groups,size(body_idx));
           else
-            pts = getTerrainContactPoints(obj.body(bi),contact_groups{i});
+            sizecheck(contact_groups,size(body_idx));
           end
-          if ~isempty(pts)
-            terrain_contact_point_struct(end+1) = struct('pts',pts,'idx',bi);
+        end
+        terrain_contact_point_struct = struct('pts',{},'idx',{});
+        for i = 1:length(body_idx)
+          bi=body_idx(i);
+          if bi ~= 1
+            if nargin < 3
+              pts = getTerrainContactPoints(obj.body(bi));
+            else
+              pts = getTerrainContactPoints(obj.body(bi),contact_groups{i});
+            end
+            if ~isempty(pts)
+              terrain_contact_point_struct(end+1) = struct('pts',pts,'idx',bi);
```

Comment:
Perhaps, but then we'd still have the overhead of looping over all the
bodies at each dt to construct a quantity that *never* changes.

On Tue, Jan 6, 2015, 3:55 PM Russ Tedrake <notifications@github.com> wrote:

> In systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m
> <https://github.com/RobotLocomotion/drake/pull/682#discussion-diff-22551225>
> :
>
> >            end
> > -          if ~isempty(pts)
> > -            terrain_contact_point_struct(end+1) = struct('pts',pts,'idx',bi);
> > +        end
> > +        terrain_contact_point_struct = struct('pts',{},'idx',{});
> > +        for i = 1:length(body_idx)
> > +          bi=body_idx(i);
> > +          if bi ~= 1
> > +            if nargin < 3
> > +              pts = getTerrainContactPoints(obj.body(bi));
> > +            else
> > +              pts = getTerrainContactPoints(obj.body(bi),contact_groups{i});
> > +            end
> > +            if ~isempty(pts)
> > +              terrain_contact_point_struct(end+1) = struct('pts',pts,'idx',bi);
>
> it would be much cleaner to have this as an element of RigidBody.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/682/files#r22551225>.
>